Rewrite Maven: Adding xml comments within a maven POM file

Is it possible to add XML comments within a POM xml file?

For example, the use case would be to detect the presence of a maven dependency, and in that case add an XML comment associated to that dependency, either inside the `dependency` TAG or as a new TAG after or before it.

I have tried to create a recipe for this purpose but I have not been able to do it at all. The recipe is as [follows](https://github.com/ferblaca/OpenRewriteRecipe/blob/main/src/main/java/com/ferblaca/openrewrite/recipes/AddCommentIntoMavenDependencyTag.java):

```
  @Override
  protected TreeVisitor<?, ExecutionContext> getVisitor() {
    return new MavenIsoVisitor<ExecutionContext>() {
      @Override
      public Xml.Tag visitTag(final Xml.Tag tag, final ExecutionContext ctx) {
        if (this.isDependencyTag("org.jasypt", "jasypt")) {

          // xml comment
          final Xml.Comment customComment = new Xml.Comment(randomId(),
              tag.getPrefix(),
              Markers.EMPTY,
              "custom comment into dependency!!!");

          // Change Content of dependency Tag
          this.doAfterVisit(new ChangeTagContentVisitor<>(tag, Collections.singletonList(customComment)));
        }
        return super.visitTag(tag, ctx);
      }
    };
  }
```

But this way I only manage to mash the content of the dependency with the comment, instead of adding it to the Content `groupId` and `artifactId`...

For a POM with this content before:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.6.5</version>
        <relativePath/>
    </parent>
    <groupId>com.example.openrewrite.test</groupId>
    <artifactId>demoOpenrewrite</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <name>demoOpenRewrite</name>
    <properties>
        <java.version>11</java.version>
    </properties>
    <dependencies>
        <dependency>
            <groupId>org.jasypt</groupId>
            <artifactId>jasypt</artifactId>
            <version>1.9.3</version>
        </dependency>
    </dependencies>
</project>
```

The result obtained after is:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.6.5</version>
        <relativePath/>
    </parent>
    <groupId>com.example.openrewrite.test</groupId>
    <artifactId>demoOpenrewrite</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <name>demoOpenRewrite</name>
    <properties>
        <java.version>11</java.version>
    </properties>
    <dependencies>
        <dependency>
            <!--custom comment into dependency!!!-->
        </dependency>
    </dependencies>
</project>
```

When the expected/desired is:

```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-parent</artifactId>
        <version>2.6.5</version>
        <relativePath/>
    </parent>
    <groupId>com.example.openrewrite.test</groupId>
    <artifactId>demoOpenrewrite</artifactId>
    <version>0.0.1-SNAPSHOT</version>
    <name>demoOpenRewrite</name>
    <properties>
        <java.version>11</java.version>
    </properties>
    <dependencies>
        <dependency>
            <!--custom comment into dependency!!!-->
            <groupId>org.jasypt</groupId>
            <artifactId>jasypt</artifactId>
            <version>1.9.3</version>
        </dependency>
    </dependencies>
</project>
```


Is it possible to do this? It would be interesting to be able to add comments, either from maven, yaml, java...

thank you very much.

hi @traceyyoshima, 

> Yes, it is possible with a slight adjustment. The ChangeTagContentVisitor does a replacement of the content in the tag through t = t.withContent(content);. So, you'll need to add the comment to the Tag's content in the position you'd like the comment.

Yes, I tried several ways to concatenate the `Xml.Comment` to the `tag.getContent()` list without success.

For the following code:

```
// Change Content of dependency Tag
this.doAfterVisit(new ChangeTagContentVisitor<>(tag, ListUtils.concat(Collections.singletonList(customComment), tag.getContent())));
```

The compilation error is as follows:

```
[ERROR] /home/fbc/workspaces/ferblaca/OpenRewriteRecipe/src/main/java/com/ferblaca/openrewrite/recipes/AddCommentIntoMavenDependencyTag.java:[37,15] cannot infer type arguments for org.openrewrite.xml.ChangeTagContentVisitor<>
[ERROR]   reason: inference variable T has incompatible bounds
[ERROR]     equality constraints: org.openrewrite.xml.tree.Content
[ERROR]     lower bounds: java.util.List<capture#1 of ? extends org.openrewrite.xml.tree.Content>,org.openrewrite.xml.tree.Xml.Comment
```

> There is a recipe that marks dependencies without modifying the AST if you'd prefer to identify the dependencies without a long-lasting comment. FindDependency will add a Marker to the dependency so that it's printed in a diff, but does not change the AST.

So it wouldn't change the AST, would it? As a suggestion, it would be nice to be able to add comments for the different types of Sources.

--- a/rewrite-maven/src/main/java/org/openrewrite/maven/AddCommentToMavenDependency.java
+++ b/rewrite-maven/src/main/java/org/openrewrite/maven/AddCommentToMavenDependency.java
@@ -1,0 +1,97 @@
+/*
+ * Copyright 2022 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.maven;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Incubating;
+import org.openrewrite.Option;
+import org.openrewrite.Recipe;
+import org.openrewrite.marker.Markers;
+import org.openrewrite.xml.XPathMatcher;
+import org.openrewrite.xml.tree.Content;
+import org.openrewrite.xml.tree.Xml;
+
+import java.util.*;
+
+import static org.openrewrite.Tree.randomId;
+
+@Incubating(since = "7.24.0")
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class AddCommentToMavenDependency extends Recipe {
+
+    @Option(displayName = "XPath",
+            description = "An XPath expression used to find matching tags.",
+            example = "/project/dependencies/dependency")
+    String xPath;
+
+    @Option(displayName = "Group",
+            description = "The first part of a dependency coordinate 'com.google.guava:guava:VERSION'.",
+            example = "com.google.guava")
+    String groupId;
+
+    @Option(displayName = "Artifact",
+            description = "The second part of a dependency coordinate 'com.google.guava:guava:VERSION'.",
+            example = "guava")
+    String artifactId;
+
+    @Option(displayName = "Comment text",
+            description = "The text to add as a comment..",
+            example = "This is excluded due to CVE <X> and will be removed when we upgrade the next version is available.")
+    String commentText;
+
+    @Override
+    public String getDisplayName() {
+        return "Add a comment to a `Maven` dependency";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Adds a comment as the first element in a `Maven` dependency.";
+    }
+
+    @Override
+    protected MavenVisitor<ExecutionContext> getVisitor() {
+        return new MavenIsoVisitor<ExecutionContext>() {
+            final XPathMatcher matcher = new XPathMatcher(xPath);
+
+            @Override
+            public Xml.Tag visitTag(Xml.Tag tag, ExecutionContext ctx) {
+                Xml.Tag t = super.visitTag(tag, ctx);
+                if (matcher.matches(getCursor()) && this.isDependencyTag(groupId, artifactId)) {
+                    if (tag.getContent() != null) {
+                        List<Content> contents = new ArrayList<>(tag.getContent());
+                        boolean containsComment = contents.stream()
+                                .anyMatch(c -> c instanceof Xml.Comment &&
+                                        commentText.equals(((Xml.Comment) c).getText()));
+                        if (!containsComment) {
+                            int insertPos = 0;
+                            Xml.Comment customComment = new Xml.Comment(randomId(),
+                                    contents.get(insertPos).getPrefix(),
+                                    Markers.EMPTY,
+                                    commentText);
+                            contents.add(insertPos, customComment);
+                            t = t.withContent(contents);
+                        }
+                    }
+                }
+                return t;
+            }
+        };
+    }
+}

--- a/rewrite-maven/src/test/kotlin/org/openrewrite/maven/AddCommentToMavenDependencyTest.kt
+++ b/rewrite-maven/src/test/kotlin/org/openrewrite/maven/AddCommentToMavenDependencyTest.kt
@@ -1,0 +1,85 @@
+/*
+ * Copyright 2022 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.maven
+
+import org.junit.jupiter.api.Test
+import org.openrewrite.test.RewriteTest
+
+class AddCommentToMavenDependencyTest : RewriteTest {
+
+    @Test
+    fun removeDependency() = rewriteRun(
+        { spec ->
+            spec.recipe(
+                AddCommentToMavenDependency(
+                    "/project/dependencies/dependency",
+                    "com.google.guava",
+                    "guava",
+                    " Comment text "
+                )
+            )
+        },
+        pomXml(
+            """
+                <project>
+                  <modelVersion>4.0.0</modelVersion>
+                  
+                  <groupId>com.mycompany.app</groupId>
+                  <artifactId>my-app</artifactId>
+                  <version>1</version>
+                  
+                  <dependencies>
+                    <dependency>
+                      <groupId>com.google.guava</groupId>
+                      <artifactId>guava</artifactId>
+                      <version>29.0-jre</version>
+                    </dependency>
+                    <dependency>
+                      <groupId>junit</groupId>
+                      <artifactId>junit</artifactId>
+                      <version>4.13.1</version>
+                      <scope>test</scope>
+                    </dependency>
+                  </dependencies>
+                </project>
+            """,
+            """
+                <project>
+                  <modelVersion>4.0.0</modelVersion>
+                  
+                  <groupId>com.mycompany.app</groupId>
+                  <artifactId>my-app</artifactId>
+                  <version>1</version>
+                  
+                  <dependencies>
+                    <dependency>
+                      <!-- Comment text -->
+                      <groupId>com.google.guava</groupId>
+                      <artifactId>guava</artifactId>
+                      <version>29.0-jre</version>
+                    </dependency>
+                    <dependency>
+                      <groupId>junit</groupId>
+                      <artifactId>junit</artifactId>
+                      <version>4.13.1</version>
+                      <scope>test</scope>
+                    </dependency>
+                  </dependencies>
+                </project>
+            """
+        )
+    )
+}

